### ISSUE TITLE
Cannot get Dorado to work on a Jetson AGX

### ISSUE THREAD

**Issue by hasindu2008** [nanoporetech/dorado#772]
# Issue Report

## Please describe the issue:

Get the following error when trying to get Dorado working on a Jetson AGX.
```
[ont/warning] dorado logger: Caught Torch error 'CUDA error: no kernel image is available for execution on the device
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.', clearing CUDA cache and retrying.
terminate called after throwing an instance of 'c10::CUDAError'
  what():  CUDA error: no kernel image is available for execution on the device
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Exception raised from launch_unrolled_kernel at /pytorch/aten/src/ATen/native/cuda/CUDALoops.cuh:132...
```

## Steps to reproduce the issue:

- using pip install ont-pyguppy-client-lib==7.2.13
- download the relavent server version
- execute basecalling on a small dataset 

## Run environment:
- Dorado version: server 7.2.13
- Dorado command: any basecalling command that uses GPU
- Operating system: Ubuntu 20.04.6 LTS
- Hardware (CPUs, Memory, GPUs): ARMv8 Processor, 16GB RAM, 512-Core NVIDIA Volta GPU (I think this is a Xavier AGX)

## Logs

```
[ont/warning] dorado logger: Caught Torch error 'CUDA error: no kernel image is available for execution on the device
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.', clearing CUDA cache and retrying.
terminate called after throwing an instance of 'c10::CUDAError'
  what():  CUDA error: no kernel image is available for execution on the device
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Exception raised from launch_unrolled_kernel at /pytorch/aten/src/ATen/native/cuda/CUDALoops.cuh:132...
```

**Comment by tijyojwad** [comment]
Hi - did you test this with the linux arm64 build? can you try the following standalone version to see if that works? https://cdn.oxfordnanoportal.com/software/analysis/dorado-0.6.1-linux-arm64.tar.gz

**Comment by hasindu2008** [comment]
Yes, that seems to work. So is it that the 7.2.13 of basecall-server is not supporting the ARM GPUs which the future ones will?

**Comment by tijyojwad** [comment]
Hi, one is available through the Nanopore community page (Linux ARM). Is that the one you installed?

![image](https://github.com/nanoporetech/dorado/assets/1127155/4ee0760d-b989-4943-bd36-b57312bdcce9)

**Comment by hasindu2008** [comment]
yes, that is the one I installed - the file name is ont-dorado-server_7.2.13_linuxaarch64_cuda10.tar.gz.

**Comment by tijyojwad** [comment]
Right, sorry for the confusion. The basecall server build is for the TX2 and based on CUDA 10. But Jetson AGX Xavier requires a more recent compute arch version for the CUDA kernels. Dorado is compiled for that, but we don't have a basecall server version released for it (and as of yet, don't have plans to support Jetson platforms beyond TX2 on basecall server)

**Comment by hasindu2008** [comment]
Oh no :( 
Will it be possible for newer ARM architectures like Xavier and Orin to be supported? Or if the basecall server source is available we could compile it for these architectures.

**Comment by neuropathbasel** [comment]
Hi there - we're also using a patched Mk1c build of MinKNOW and would be highly interested in getting a working dorado basecall server for the Xavier and ORIN. Previous versions of guppy were working well (and are still in use) on these devices. Furtermore, the aarch64 build of dorado (standalone) works well on the ORIN. Would you be so kind as to maintain an aarch64 branch of MinKNOW/dorado? There are many advantages of embedded devices like the Xavier/ORIN in medical diagnostics with respect to registering as laboratory-developed test. See [https://actaneurocomms.biomedcentral.com/articles/10.1186/s40478-024-01759-2](https://actaneurocomms.biomedcentral.com/articles/10.1186/s40478-024-01759-2). The best would be to distribute sources of the basecall server functionality, it would be most flexible since differnt diagnostic and scientific scenarios might profit from different SoC/SoM computers. Thanks!

**Comment by tijyojwad** [comment]
thanks for the inputs @hasindu2008 and @neuropathbasel - we'll discuss internally and get back to you

**Comment by hasindu2008** [comment]
@tijyojwad Thank you very much for your consideration. In addition, I really would like to ask ONT to consider maintaining support for ARM platforms like Jetson, not only for basecall servers but also minknow for the future. ARM devices are affordable devices and having support on such low-cost systems would be quite beneficial for small labs, and researchers at developing countries. I think that support for such devices is very much inline with ONT's vision for democratising genomics.

**Comment by neuropathbasel** [comment]
@tijyojwad Thanks from our side as well. We would like to briefly add that many regions of the world are unlikely to get their patients access to diagnostics based on classical (costly) sequencing platforms. [NanoDiP ](https://actaneurocomms.biomedcentral.com/articles/10.1186/s40478-024-01759-2) is particularly suited for those scenarios. In addition, remote support for relatively well-defined compute platforms (such as the jetsons) is more realistic than supporting a mulitude of PC/Mac etc. platforms. A long-term support ARM release of MinKNOW is a pre-requisite to get things into medical diagnostics.

**Comment by MatDaw** [comment]
Hello, I'm also having issues with installing minknow and dorado basecall server on Jetson AGX Orin. I know it's not officially supported, but I've been following instructions from [Miles Benton](https://github.com/sirselim/jetson_nanopore_sequencing/blob/main/setup-guide-mk1c.txt) and [David Eccles](https://gringer.gitlab.io/presentation-notes/2021/10/08/gpu-calling-in-minknow/) and got same CUDA error as the one in this topic.
From what I've gathered we should not be using latest versions of software, so is there a way to download specific previous versions? From what I've found we can download previous versions of [dorado](https://github.com/nanoporetech/dorado/releases) and [guppy](https://pypi.org/project/ont-pyguppy-client-lib/#history), but cannot find minknow and dorado_basecall_server. Listing apt only shows latest version available for download:
![image](https://github.com/user-attachments/assets/c9db9859-67ba-4232-9159-9e0c7329e34e).
Would appreciate any help!

**Comment by neuropathbasel** [comment]
Hi!

The only version of dorado working on the ORIN is the standalone binary, e.g., version 5.x. This implies that you can't use it from MinKNOW for live basecalling. The GPU is different between the MK1C and the ORIN. Nandip has been adapted to perform only batch-wise post-hoc basecalling aling with alignment and methylation calling. This is unfortunate, but the only people who could solve this are at ONT.

Sorry for the bad news.

Best

J.


________________________________
Von: matdawidziuk ***@***.***>
Gesendet: Montag, 12. August 2024 20:42
An: nanoporetech/dorado
Cc: Hench Juergen, Dr.; Mention
Betreff: Re: [nanoporetech/dorado] Cannot get Dorado to work on a Jetson AGX (Issue #772)

Hello, I'm also having issues with installing minknow and dorado basecall server on Jetson AGX Orin. I know it's not officially supported, but I've been following instructions from Miles Benton and David Eccles and got same CUDA error as the
ZjQcmQRYFpfptBannerStart
Diese Nachricht stammt von einem externen Absender
Diese Nachricht kam von ausserhalb Ihrer Organisation.

ZjQcmQRYFpfptBannerEnd

Hello, I'm also having issues with installing minknow and dorado basecall server on Jetson AGX Orin. I know it's not officially supported, but I've been following instructions from Miles Benton<https://urldefense.com/v3/__https://github.com/sirselim/jetson_nanopore_sequencing/blob/main/setup-guide-mk1c.txt__;!!EDSXHx-qqdzzoNk!ufMgVYBPn4-bLjim5LEPjTRg4UQfOGCgbE7zSPUdcfBqh8eg-rPyIAlyyAln2Ohc3HiLO3XSH96AzJe4csf6e8gmnZ0$> and David Eccles<https://urldefense.com/v3/__https://gringer.gitlab.io/presentation-notes/2021/10/08/gpu-calling-in-minknow/__;!!EDSXHx-qqdzzoNk!ufMgVYBPn4-bLjim5LEPjTRg4UQfOGCgbE7zSPUdcfBqh8eg-rPyIAlyyAln2Ohc3HiLO3XSH96AzJe4csf6WUjVZgQ$> and got same CUDA error as the one in this topic.
From what I've gathered we should not be using latest versions of software, so is there a way to download specific previous versions? From what I've found we can download previous versions of dorado<https://urldefense.com/v3/__https://github.com/nanoporetech/dorado/releases__;!!EDSXHx-qqdzzoNk!ufMgVYBPn4-bLjim5LEPjTRg4UQfOGCgbE7zSPUdcfBqh8eg-rPyIAlyyAln2Ohc3HiLO3XSH96AzJe4csf6iL1Erxc$> and guppy<https://urldefense.com/v3/__https://pypi.org/project/ont-pyguppy-client-lib/*history__;Iw!!EDSXHx-qqdzzoNk!ufMgVYBPn4-bLjim5LEPjTRg4UQfOGCgbE7zSPUdcfBqh8eg-rPyIAlyyAln2Ohc3HiLO3XSH96AzJe4csf6Fa5KELM$>, but cannot find minknow and dorado_basecall_server. Listing apt only shows latest version available for download:
image.png (view on web)<https://urldefense.com/v3/__https://github.com/user-attachments/assets/c9db9859-67ba-4232-9159-9e0c7329e34e__;!!EDSXHx-qqdzzoNk!ufMgVYBPn4-bLjim5LEPjTRg4UQfOGCgbE7zSPUdcfBqh8eg-rPyIAlyyAln2Ohc3HiLO3XSH96AzJe4csf6J-iRGnI$>.
Would appreciate any help!

**Comment by malton-ont** [comment]
Hi all,

Dorado basecall server 7.11.2 has just been released, and provides an Orin-compatible build. This is included in Minknow 25.09.
